### PR TITLE
Add support for Next.js basePath

### PR DESCRIPTION
**Pull request by alexkirsz**
# Add support for Next.js basePath

This PR adds support for the `basePath` next.config.js option. My first attempt at this was to change the `RouterContentSource` so it could strip a base path prefix from all routes, so underlying content sources did not need to know any concept of base path, and add that prefix to all file names exposed to the client via Next.js' assetPrefix. However, this breaks down quickly since we still need some knowledge of the server-root-relative path in all assets that load other assets.

This PR passes down a `server_path` alongside the `server_root` when creating sources. `server_root` is the server root (e.g. ``) while `server_path` is the server root + any base path or asset prefix (e.g. `/basepath`).

### Issue reporting

The issues currently look terrible. I'm not sure what's going on here, but emitting an issue from `validate_next_config` seems to work poorly:

image

Furthermore, Next.js exits on startup when the `basePath` config key is invalid. In order to reproduce this behavior accurately, we would need to load next.config.js eagerly when starting the dev server, instead of lazily as is currently done.

### TODO
* Full `assetPrefix` support. Right now we always use basePath instead, but they can differ. I've created an issue for this.

### CODE REVIEW

File: crates/next-core/src/app_source.rs [hunk 1]
```diff
@@ -112,7 +114,7 @@ async fn next_client_transition(
 
     Ok(NextClientTransition {
         is_app: true,
-        server_root,
+        server_root: server_path,
```
Comment (alexkirsz) [inline crates/next-core/src/app_source.rs:117]: @sokra I'm not sure this is correct. Perhaps this needs both `server_root` and `server_path` to be passed down?

File: crates/next-core/src/server_rendered_source.rs [hunk 2]
```diff
@@ -112,7 +112,7 @@ pub async fn create_server_rendered_source(
         client_module_options_context,
         client_resolve_options_context,
         client_environment,
-        server_root,
+        server_root: server_path,
```
Comment (alexkirsz) [inline crates/next-core/src/server_rendered_source.rs:115]: @sokra Same question.

File: crates/next-core/src/app_source.rs [hunk 3]
```diff
@@ -73,10 +71,12 @@ fn next_client_chunks_transition(
     execution_context: ExecutionContextVc,
     app_dir: FileSystemPathVc,
     server_root: FileSystemPathVc,
+    server_path: FileSystemPathVc,
     browserslist_query: &str,
 ) -> TransitionVc {
```
Comment (alexkirsz) [inline crates/next-core/src/app_source.rs:76]: @sokra Also wondering if this needs both `server_root` and `server_path`

**Review by sokra — APPROVED** [review]

**Review by jridgewell — CHANGES_REQUESTED** [review]
Still trying to review, but this is breakig static asset serving the `/public` folder.

**Review by sokra — CHANGES_REQUESTED** [review]
I want to block this on the benchmark results. A slowdown would be acceptable, but it's weird that such a slowdown happens on that kind of change, which shouldn't introduce any performance cost. So there is something potentially weird here...

File: crates/turbopack-dev-server/src/source/router.rs [hunk 4]
```diff
@@ -9,28 +12,33 @@ use crate::source::ContentSourcesVc;
...
 #[turbo_tasks::value(shared)]
 pub struct RouterContentSource {
+    pub base_path: OptionStringVc,
     pub routes: Vec<(String, ContentSourceVc)>,
     pub fallback: ContentSourceVc,
 }
 
 impl RouterContentSource {
-    fn get_source<'s, 'a>(&'s self, path: &'a str) -> (&'s ContentSourceVc, &'a str) {
+    async fn get_source<'s, 'a>(&'s self, path: &'a str) -> Result<(&'s ContentSourceVc, &'a str)> {
```
Comment (jridgewell) [inline crates/turbopack-dev-server/src/source/router.rs:21]: Nit: Why was this made async?

File: crates/turbopack-core/src/resolve/options.rs [hunk 5]
```diff
@@ -6,7 +6,10 @@ use turbo_tasks::{
     debug::ValueDebugFormat, primitives::StringVc, trace::TraceRawVcs, TryJoinIterExt, Value,
     ValueToString, ValueToStringVc,
 };
-use turbo_tasks_fs::{glob::GlobVc, FileSystemPathVc};
+use turbo_tasks_fs::{
+    glob::{Glob, GlobVc},
+    FileSystemPathVc,
+};
```
Comment (jridgewell) [inline crates/turbopack-core/src/resolve/options.rs:12]: Rouge edit? Clippy is complaining about unused `Glob`.

```suggestion
use turbo_tasks_fs::{glob::GlobVc, FileSystemPathVc};
```

**Review by jridgewell — APPROVED** [review]
Static asset serving is fixed, unblocking while I finish review.

File: crates/turbopack-dev-server/src/source/router.rs (hunk 4 above)
Comment (alexkirsz) [reply crates/turbopack-dev-server/src/source/router.rs:21]: Previous version used `base_path`, no longer needed.

**Comment by alexkirsz** [comment]
@sokra Through process of elimination, I can't seem to pinpoint exactly *where* this 5-10% slow down is coming from, which makes me think it's a combination of multiple changes in this PR.

What I've tried:
* Stop reading basePath from `next.config.js` entirely;
* Removing the additional argument to `get_client_chunking_context`;
* Reverting all changes to `crates/next-core/js/*`;
* Reverting changes to the env;
* And finally, changes to each file separately (depending on what's possible).

**Pull request closed without merging**